services/src/handlers/datasets.rs

```diff
@@ -38,63 +38,36 @@ use geoengine_operators::{
     util::gdal::{gdal_open_dataset, gdal_open_dataset_ex},
 };
 use snafu::ResultExt;
-use uuid::Uuid;
-use warp::Filter;
-
-pub(crate) fn list_providers_handler<C: Context>(
-    ctx: C,
-) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
-    warp::path("providers")
-        .and(warp::get())
-        .and(authenticate(ctx.clone()))
-        .and(warp::any().map(move || ctx.clone()))
-        .and(warp::query())
-        .and_then(list_providers)
-}
 
-// TODO: move into handler once async closures are available?
-async fn list_providers<C: Context>(
+pub(crate) async fn list_providers_handler<C: Context>(
     session: C::Session,
-    ctx: C,
-    options: DatasetProviderListOptions,
-) -> Result<impl warp::Reply, warp::Rejection> {
+    ctx: web::Data<C>,
+    options: web::Query<DatasetProviderListOptions>,
+) -> Result<impl Responder> {
     let list = ctx
         .dataset_db_ref()
         .await
-        .list_dataset_providers(&session, options.validated()?)
+        .list_dataset_providers(&session, options.into_inner().validated()?)
```

Comment:
Every type in the parameters of a request handler must implement the [FromRequest](https://actix.rs/docs/extractors/) trait. This tells Actix how to construct an instance of that type based on the incoming HTTP request.
Various wrappers provided by Actix can be used to handle types which do not implement `FromRequest` themselves, for example `web::Json`, `web::Query` and `web::Path`. These handle the most common tasks like parsing the request body as JSON, deserializing query parameters and extracting path segments (which were marked with `{...}` in the routing table), respectively. This means that only in more special cases like sessions the trait must be implemented directly.
If an owned reference to the data inside an Actix provided wrapper is needed, `.into_inner()` can be used, which is the same as `.0`. When only a borrow is needed, it is not necessary to use `.into_inner()` because the wrappers implement the `Deref` trait.

In the case in the comment `validated` takes `self` (and not `&self`), which means that `.into_inner()` must be called to own the inner `DatasetProviderListOptions` of `web::Query<DatasetProviderListOptions>.`